swiftsimio/initial_conditions/generate_particles.py

```diff
@@ -1,5 +1,1304 @@
+
+            xr = np.zeros((1, 3), dtype=np.float)
+            for d in range(ndim):
+                xr[0, d] = np.random.uniform(low=0.0, high=coord_threshold[d])
+
+            if np.random.uniform() <= rhofunc(xr, ndim) / self.rho_max:
+                x[keep] = xr
+                keep += 1
+
+        return unyt.unyt_array(x, self.unitsys["length"])
+
+    def compute_h_and_rho(
+        self, nngb: Union[int, None] = None, boxsize: Union[np.ndarray, None] = None
+    ):
+        r"""
+        Compute actual smoothing lengths and particle densities.
+
+        Parameters
+        -----------------
+
+        nngb: integer or None, optional
+            number of neighbours to search for for each particle. If ``None``,
+            ``self.iterparams.Nngb_int``  will be used, which is
+            the automatically set up value for given parameters during
+            initialisation.
+
+        boxsize: np.ndarray or None, optional
+            boxsize to be used to generate the tree. If ``None``,
+            ``self.iterparams.boxsize_for_tree`` will be used, which is
+            the automatically set up value for given parameters during
+            initialisation.
+
+
+        Returns
+        --------------
+
+        h: np.ndarray
+            numpy array of particle smoothing lengths with shape (``self.npart``)
+
+        rho: np.ndarray
+            numpy array of particle densities with shape (``self.npart``)
+
+        """
+        if nngb is None:
+            nngb = self.iterparams.Nngb_int
+        if boxsize is None:
+            boxsize = self.iterparams.boxsize_for_tree
+
+        ndim = self.ndim
+        kernel_func, _, kernel_gamma = get_kernel_data(self.kernel, ndim)
+
+        rho = np.zeros(self.npart, dtype=np.float)
+        h = np.zeros(self.npart, dtype=np.float)
+
+        tree = KDTree(self.x[:, :ndim], boxsize=boxsize)
+        for p in range(self.npart):
+            dist, neighs = tree.query(self.x[p, :ndim], k=nngb)
+            h[p] = dist[-1] / kernel_gamma
+            for i, n in enumerate(neighs):
+                W = kernel_func(dist[i], dist[-1])
+                rho[p] += W * self.m[n]
+
+        return h, rho
+
+    def iteration_step(
+        self, iteration: int,
+    ):
+        """
+        Perform one step of the actual iteration.
+
+        Parameters
+        -------------
+
+        iteration: int
+            current iteration number
+
+
+        Returns
+        --------------
+
+        converged: bool
+            Whether the iteration satisfies convergence criteria.
+
+        """
+        ndim = self.ndim
+        periodic = self.periodic
+        boxsize = self.boxsize_to_use
+        npart = self.npart
+        rhofunc = self.rhofunc
+        x = self.x
+        runparams = self.runparams
+        ipars = self.iterparams
+
+        from .IC_plotting import IC_plot_current_situation
+
+        # kernel data
+        kernel_func, _, kernel_gamma = get_kernel_data(self.kernel, ndim)
```

Comment:
Unused parameter `kernel_gamma`, is that intended?

Reply:
Yes, just to mess with you.